doc/libvmmalloc.3

```diff
@@ -0,0 +1,229 @@
+.\"       notice, this list of conditions and the following disclaimer in
+.\"       the documentation and/or other materials provided with the
+.\"       distribution.
+.\"
+.\"     * Neither the name of Intel Corporation nor the names of its
+.\"       contributors may be used to endorse or promote products derived
+.\"       from this software without specific prior written permission.
+.\"
+.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+.\"
+.\"
+.\" libvmmalloc.3 -- man page for libvmmalloc
+.\"
+.\" Format this man page with:
+.\"	man -l libvmmalloc.3
+.\" or
+.\"	groff -man -Tascii libvmmalloc.3
+.\"
+.TH libvmmalloc 3 "vmmalloc API version 0.1" "NVM Library"
+.SH NAME
+libvmmalloc \- general purpose volatile memory allocation library
+.SH SYNOPSIS
+.PP
+.sp
+.ft B
+.nf
+#include <stdlib\&.h>
+#include <malloc\&.h>
+#include <libvmmalloc\&.h>
+.fi
+.ft
+.HP \w'void\ *malloc('u
+.BI "void *malloc(size_t\ " "size" ");"
+.HP \w'void\ *calloc('u
+.BI "void *calloc(size_t\ " "number" ", size_t\ " "size" ");"
+.HP \w'int\ posix_memalign('u
+.BI "int posix_memalign(void\ **" "memptr" ", size_t\ " "alignment" ", size_t\ " "size" ");"
+.HP \w'void\ *realloc('u
+.BI "void *realloc(void\ *" "ptr" ", size_t\ " "size" ");"
+.HP \w'void\ free('u
+.BI "void free(void\ *" "ptr" ");"
+.HP \w'size_t\ malloc_usable_size('u
+.BI "size_t malloc_usable_size(const\ void\ *" "ptr" ");"
+.HP \w'void\ *memalign('u
+.BI "void *memalign(size_t\ " "alignment" ", size_t\ " "size" ");"
+.HP \w'void\ *aligned_alloc('u
+.BI "void *aligned_alloc(size_t\ " "alignment" ", size_t\ " "size" ");"
+.HP \w'void\ *valloc('u
+.BI "void *valloc(size_t\ " "size" ");"
+.HP \w'void\ *pvalloc('u
+.BI "void *pvalloc(size_t\ " "size" ");"
+.HP \w'void\ cfree('u
+.BI "void cfree(void\ *" "ptr" ");"
+.SH DESCRIPTION
+.B libvmmalloc
+provides a set of standard system memory allocation routines, as defined in
+.BR malloc (3),
+.BR posix_memalign (3)
+and
+.BR malloc_usable_size (3),
+but working in a way similar to
+.BR libvmem (3).
+.PP
+The typical method to use
+.B libvmmalloc
+is to load it before all other libraries by setting the environment
+variable
+.BR LD_PRELOAD .
+See
+.BR ld.so (8)
+for more details.
```

Comment:
Krzysztof,

This looks great.  My only comment is to re-organize the man page a little bit.  Right now it is a bit difficult for someone unfamiliar with NVML to understand what this library is for.  Reading the man page it seems to provide a bunch of existing interfaces and points the reader to the libvmem and ld.so man pages.

Instead, I would have the SYNOPSIS section show exactly how to set LD_PRELOAD and run a program with libvmmalloc interposed.  The first sentence of the DESCRIPTION should make it clear that this library transparently converts DRAM allocations to pmem allocations and only then you should mention that in fact it works for allocating from any memory-mapped file.  And then you put a big bold caveat on the type of programs where this won't work (like those that fork and don't exec).

Finally, one nit: update your copyright years so Lukasz doesn't nag you about it :-)